### PR TITLE
infoschema: don't load table info to get auto_increment value

### LINKED ISSUE #57295: Query to `is.tables` is slow in infoschema cache v2
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Create a TiDB cluster that has a lot of schemas and tables.
2. Execute `select * from information_schema.tables where table_name = 'xx'`

### 2. What did you expect to see? (Required)

It should complete fast.

### 3. What did you see instead (Required)

It takes about 400ms to finish a query. From the CPU profile, we can see there are many cache misses.

image

### 4. What is your TiDB version? (Required)

e16613d0b

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #57295

Problem Summary:

### What changed and how does it work?

- Add a method `TableIsCached()` for `infoschemaV2`.
- Don't load table info in `getAutoIncrementID()`. If we do, the current auto ID must be zero. Let's return zero directly.
- Add a simple unit test.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  After this PR, the same query only takes 80ms.
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/infoschema/test/infoschemav2test/v2_test.go [hunk 1]
```diff
@@ -534,3 +535,26 @@ func TestSnapshotInfoschemaReader(t *testing.T) {
...
+	tk.MustQuery(autoIncQuery).Check(testkit.Rows("0"))
+	tk.MustExec("insert into t values (),(),();")
+	tk.MustQuery(autoIncQuery).Check(testkit.Rows("4"))
+
+	tk.MustExec("set @@global.tidb_schema_cache_size = 1024 * 1024 * 1024;")
+	tk.MustExec("create table t1 (a int);") // trigger infoschema cache reload
+	tk.MustQuery(autoIncQuery).Check(testkit.Rows("0"))
+	tk.MustExec("insert into t values ();")
+	tk.MustQuery(autoIncQuery).Check(testkit.Rows("5"))
+	tk.MustExec("set @@global.tidb_schema_cache_size = 0;")
```
Comment (tiancaiamao) [inline pkg/infoschema/test/infoschemav2test/v2_test.go:557]: There is an API you can use to evict table cache directly, although trigger reload can get the same effect.

```
	is := dom.InfoSchema()
	ok, v2 := infoschema.IsV2(is)
	require.True(t, ok)
	v2.EvictTable(model.NewCIStr("test"), model.NewCIStr("t1"))
```

**Review by tiancaiamao — APPROVED** [review]

File: pkg/executor/infoschema_reader.go [hunk 2]
```diff
@@ -242,11 +242,29 @@ func getAutoIncrementID(
 	sctx sessionctx.Context,
 	tblInfo *model.TableInfo,
 ) int64 {
+	if raw, ok := is.(*infoschema.SessionExtendedInfoSchema); ok {
```
Comment (CbcWestwolf) [inline pkg/executor/infoschema_reader.go:245]: Why do we need this type assertion?

Comment (tangenta) [reply pkg/executor/infoschema_reader.go:245]: `Infoschema` that we can get from session is wrapped with `SessionExtendedInfoSchema`:
 https://github.com/pingcap/tidb/blob/a5832b4689acf925173227ccd59d06a5acd75e5c/pkg/table/temptable/infoschema.go#L23

File: pkg/infoschema/test/infoschemav2test/v2_test.go (hunk 1 above)
Comment (tangenta) [reply pkg/infoschema/test/infoschemav2test/v2_test.go:557]: The schema is not v2 if we don't reload.

**Review by CbcWestwolf — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**